src/main/java/com/endava/internship/collections/StudentMap.java

```diff
@@ -1,72 +1,396 @@
+        System.out.println("resize");
+        Node[] oldTable = table;
+        size = 0;
+        int n = Math.min(oldTable.length + oldTable.length, MAXIMUM_CAPACITY);
+        table = new Node[n];
+        for(int i = 0; i < oldTable.length; i++) {
+            if(oldTable[i] != null) {
+                while(oldTable[i].getNext() != null) {
+                    Node lastNode = oldTable[i];
+                    while(lastNode.getNext().getNext() != null) {
+                        lastNode = lastNode.getNext();
+                    }
+                    putNode(lastNode.getNext());
+                    lastNode.setNext(null);
+                }
+                putNode(oldTable[i]);
+                oldTable[i] = null;
+            }
+        }
+    }
     @Override
     public int size() {
-        //TODO
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        //TODO
-        return false;
+        return size == 0;
     }
 
     @Override
     public boolean containsKey(Object o) {
-        //TODO
+        if(o instanceof Student || o == null) {
+            Node node = getNode(new Node(hash(((Student) o)), ((Student) o), null, null));
+            return node != null;
+        }
         return false;
     }
 
     @Override
     public boolean containsValue(Object o) {
-        //TODO
+        if(table != null && size > 0 &&
+                (o instanceof Integer || o == null)) {
+            for(int i = 0; i < table.length; i++) {
+                for(Node node = table[i]; node != null; node = node.getNext()) {
+                    if(node.getVal() == o || node.getVal().equals(o)) {
+                        return true;
+                    }
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public Integer get(Object o) {
-        //TODO
+        if(o instanceof Student || o == null) {
+            Node node = getNode(new Node(hash(((Student) o)), ((Student) o), null, null));
+            return node == null ? null : node.getVal();
+        }
+        return null;
+    }
+    private Node getNode(Node node) {
+        if(table != null && size > 0) {
+            Node listNode = node.getKey() == null ? table[0] :
+                    table[(table.length - 1) & node.getHash()];
+            if(listNode != null) {
+                while(true) {
+                    if(listNode.equals(node)) {
+                        node.setVal(listNode.getVal());
+                        return node;
+                    }
+                    else {
+                        if(listNode.getNext() == null) {
+                            break;
+                        }
+                        else {
+                            listNode = listNode.getNext();
+                        }
+                    }
+                }
+            }
+        }
         return null;
     }
-
     @Override
     public Integer put(Student student, Integer integer) {
```

Comment:
please verify these method, seems somewhere here is a bug. I tried to put in your hashmap more than 20 objects, and I got an exception

Reply:
fixed